Ab/implement multi base custom reader

abarciauskas-bgse opened the pull request:
This PR:
* removes the items endpoints since we don't need them yet
* adds the AssetReader class which inherits from MultiBaseReader and enables dynamic definition and use of the asset reader depending on the asset type. Mostly a copy paste of the implementation of the [tile method for the MultiBaseReader](https://github.com/cogeotiff/rio-tiler/pull/711/files#diff-aee02035a0c41599c6860bf04d8075994bb3f9e5b1d5d0595c0d02c5f4ddf596L455-L543) in https://github.com/cogeotiff/rio-tiler/pull/711. The XarrayReader is not implemented yet so the _get_reader method will throw an error for the hdf5/netcdf media types
* Adds a few tests

A few questions / todos which may turn into separate issues:
- [ ] would like to add a tests for index parameters since its a bit of unclear how that is intended to work in the
- [ ] fix test for tile method in asset reader (very slow)
- [ ] I noticed the AssetReader, since it depends on the MultiBaseReader and thus the spatial mixin requires input.bbox and input.assets - is there a way to include this in the class attributes so it's clear these keys in the input dict are required?

abarciauskas-bgse commented on titiler/stacapi/asset_reader.py:
```diff
@@ -0,0 +1,222 @@
+            info["type"] = asset_info["type"]
+
+        if header_size := asset_info.get("file:header_size"):
+            info["env"]["GDAL_INGESTED_BYTES_AT_OPEN"] = header_size  # type: ignore
```
it may be worth asking MS PC folks if the file extension or statistics object of the raster extension are ever employed, because if they're never used this code and the block below are unnecessary.

abarciauskas-bgse replied:
same goes for alternate URL

vincentsarago replied:
it doesn't hurt to keep this

abarciauskas-bgse commented on titiler/stacapi/asset_reader.py:
```diff
@@ -0,0 +1,222 @@
+        asset_indexes = asset_indexes or {}
+
+        # We fall back to `indexes` if provided
+        indexes = kwargs.pop("indexes", None)
```
Not sure where indexes come from or how they are used

vincentsarago replied:
indexes comes from the `bidx` query-parameter, but for asset based backend we usually use `asset_bidx` option. We allow both usually

vincentsarago commented on titiler/stacapi/backend.py:
```diff
@@ -128,8 +45,8 @@ class STACAPIBackend(BaseBackend):
-    # Use Custom STAC reader (outside init)
-    reader: Type[CustomSTACReader] = attr.ib(init=False, default=CustomSTACReader)
+    # Use custom asset reader (outside init)
+    reader: Type[AssetReader] = attr.ib(init=False, default=AssetReader)
```
~~Just be careful, the reader we use within the `STACAPIBackend` should accept a simple `Dict` object~~

https://github.com/developmentseed/titiler-stacapi-mspc/blob/ab/implement-multi-base-custom-reader/titiler/stacapi/backend.py#L214

~~which is why we had a small `CustomSTACReader` with `input: Dict[str, Any] = attr.ib()` as input~~

~~Note: in the default STACReader, the input is either a string or pystac object, but here we create the Dict directly form the pgstac query~~

edit: I see this is done in AssetReader 😅

vincentsarago commented on titiler/stacapi/asset_reader.py:
```diff
@@ -0,0 +1,222 @@
+                    asset_info["url"], tms=self.tms, **self.reader_options
+                ) as src:
+                    if type(src) == XarrayReader:
+                        raise NotImplementedError("XarrayReader not yet implemented")
```
I don't think we need to do this here, at least we can raise an issue in `_get_reader` for now if we want

abarciauskas-bgse commented on tests/test_asset_reader.py:
```diff
@@ -0,0 +1,62 @@
+    empty_stac_reader._get_reader(asset_info)
+
+
+@pytest.mark.skip(reason="Too slow.")
```
this is very slow for some reason, even though as far as I can tell the mock_rasterio_open is working as expected.

abarciauskas-bgse commented on titiler/stacapi/factory.py:
```diff
@@ -195,35 +194,32 @@ def tile(
             scale = scale or 1
 
             tms = self.supported_tms.get(tileMatrixSetId)
-            with rasterio.Env(**env):
```
I removed this context manager because I saw one was being used in AssetReader#tile#_reader `with self.ctx` but should we have it here as well since as you mentioned there are 3 levels of reader (STAC, item and asset)?

abarciauskas-bgse replied:
I reverted this change as I could see why different context manager blocks are needed since the `env` passed in the request could be different than the env passed in the asset_info, but I would be curious to know it's intentional to have multiple nested rasterio.Env context managers.

vincentsarago replied:
it's intentional to have an env for each reader, so we can set variable like `GDAL_INGESTED_BYTES_AT_OPEN` using specific number

abarciauskas-bgse commented:
@vincentsarago I believe I have addressed all comments, do you mind taking another look? 

There are some outstanding issues mostly around testing and type checking that I listed in the description that I will turn into issues.

vincentsarago commented on titiler/stacapi/asset_reader.py:
```diff
@@ -0,0 +1,219 @@
+                with reader(
+                    asset_info["url"], tms=self.tms, **self.reader_options
+                ) as src:
+                    data = src.tile(*args, indexes=idx, **kwargs)
```
I forgot about this but for XarrayReader we don't support the `indexes=` option

abarciauskas-bgse replied:
ok so do you think I should remove this even though we're not implementing the xarrayreader yet?

vincentsarago replied:
no we still need to be able to handle indexes for rasterio reader, so we need to do something like 

```python
if idx is not None:
    kwargs.update({"indexes": idx})
```
